numpy/__init__.pyi

```diff
@@ -486,16 +493,352 @@ where: Any
 who: Any
 
 _NdArraySubClass = TypeVar("_NdArraySubClass", bound=ndarray)
+_DTypeScalar = TypeVar("_DTypeScalar", bound=generic)
 _ByteOrder = Literal["S", "<", ">", "=", "|", "L", "B", "N", "I"]
 
-class dtype:
+class dtype(Generic[_DTypeScalar]):
     names: Optional[Tuple[str, ...]]
-    def __init__(
-        self,
-        dtype: DtypeLike,
+    # Overload for subclass of generic
+    @overload
+    def __new__(
+        cls,
+        dtype: Type[_DTypeScalar],
         align: bool = ...,
         copy: bool = ...,
-    ) -> None: ...
+    ) -> dtype[_DTypeScalar]: ...
+    # Overloads for string aliases, Python types, and some assorted
+    # other special cases. Order is sometimes important because of the
+    # subtype relationships
+    #
+    # bool < int < float < complex
+    #
+    # so we have to make sure the overloads for the narrowest type is
+    # first.
+    @overload
+    def __new__(
+        cls,
+        dtype: Union[
+            Type[bool],
+            Literal[
+                "?",
+                "=?",
+                "<?",
+                ">?",
+                "bool",
+                "bool_",
+            ],
+        ],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[bool_]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal[
+            "uint8",
+            "u1",
+            "=u1",
+            "<u1",
+            ">u1",
+            "B",
+            "=B",
+            "<B",
+            ">B",
+        ],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[uint8]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal[
+            "uint16",
+            "u2",
+            "=u2",
+            "<u2",
+            ">u2",
+            "h",
+            "=h",
+            "<h",
+            ">h",
+        ],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[uint16]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal[
+            "uint32",
+            "u4",
+            "=u4",
+            "<u4",
+            ">u4",
+            "I",
+            "=I",
+            "<I",
+            ">I",
+        ],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[uint32]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal[
+            "uint64",
+            "u8",
+            "=u8",
+            "<u8",
+            ">u8",
+            "L",
```

Comment:
I'd leave out ``"L"``- and ``"l"``-based char codes for now, 
as those can correspond to either 64- or 32-bit integers depending on the platform.

Reply:
Can you give a reference for that? This is surprising given that

```
In [5]: np.dtype('int64').char
Out[5]: 'l'

In [6]: np.dtype('int32').char
Out[6]: 'i'
```

Reply:
It's a maybe not as explicit as it should be, but the [generic docs](https://numpy.org/doc/stable/reference/arrays.scalars.html#built-in-scalar-types) do mention platform specifics 
(these `compatible: ...` statements). A given character code will always correspond to the 
same platform-independent type/alias (`byte`, `short`, `intc`, _etc._), but the same does not 
hold for their associated numerical precision.

On windows for example:
```python
In [1]: import numpy as np

In [2]: np.dtype('int64').char
Out[2]: 'q'

In [3]: np.dtype('int32').char
Out[3]: 'l'
```

Reply:
Ok, ditched the character codes for all int types; since they all correspond to C types they are all platform dependent. (Or at least the C standard allows them to be platform independent though they might not be in practice.)